Runtime Storage (#38)
substrate-developer-hub/knowledgebase: current/runtime/storage.md

title: Runtime Storage
---

Runtime storage allows you to store data in your blockchain that is persisted between blocks and can be accessed from
within your runtime logic. Storage should be one of the most critical concerns of a blockchain runtime developer. This
statement is somewhat self-evident, since one of the primary objectives of a blockchain is to provide decentralized
consensus about the state of the underlying storage. Furthermore, well designed storage systems reduce the load on
nodes in the network, which will lower the overhead for participants in your blockchain. Substrate exposes a set of
layered, modular storage APIs that allow runtime developers to make the storage decisions that suit them best. However,

> the fundamental principal of blockchain runtime storage is to minimize its use. This document is intended to provide

Reviewer: I think this should be principle by definition 1c
https://www.merriam-webster.com/dictionary/principle

information and best practices about Substrate's runtime storage interfaces. Please refer to
[the advanced storage documentation](../advanced/storage) for more information about how these interfaces are implemented.

## Storage Items

> [The `storage` module in the Substrate Support pallet](https://substrate.dev/rustdocs/master/frame_support/storage/index.html)

Reviewer: Support is not a pallet
```suggestion
The `storage` module in [FRAME Support](https://substrate.dev/rustdocs/master/frame_support/storage/index.html)
```

> gives runtime developers access to Substrate's flexible storage APIs. Any value which can be encoded by the

Reviewer: ```suggestion
gives runtime developers access to Substrate's flexible storage APIs. Any value that can be encoded by the
```

[Parity SCALE codec](../advanced/codec) is supported by these storage APIs:

* [Storage Value](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html) - A single value
* [Storage Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html) - A key-value hash map
* [Storage Double Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html) - An
  implementation of a map with two keys that provides the important ability to efficiently remove all entries that have
  a common first key
* Iterable Storage Maps - Implementations of
  [Storage Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageMap.html) and
  [Storage Double Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageDoubleMap.html)
  whose keys and values can be iterated over

The type of Storage Item you select should depend on the logical way in which the value will be used by your runtime.

### Storage Value

This type of storage item should be used for values that are viewed as a single unit by the runtime, whether that is a

> single primitive value, a single `struct` or a single collection of related items. Although wrapping related items in

Reviewer: ```suggestion
single primitive value, a single `struct`, or a single collection of related items. Although wrapping related items in
```

[a comprehensive list of the methods that Storage Values expose](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#required-methods).
Some of the most important methods are summarized here:

* [`get()`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#tymethod.get) - Load
  the value from storage.
* [`put(val)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#tymethod.put) - Store
  the provided value.
* [`mutate(fn)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#tymethod.mutate) -
  Mutate the value with the provided function.
* [`take()`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#tymethod.take) - Remove
  the value from storage.

### Storage Maps

Storage Maps are implemented as maps with hashed keys, which is a pattern that should be familiar to most developers.
Unlike traditional hash maps, though, [Storage Maps in Substrate are simple key-value stores](../advanced/storage) that
do not take key collision into account. The hashing algorithms that Substrate supplies are designed so that runtime

> developers do not need to worry about key collisions, but the implementation of Substrate Storage Maps does becomes

Reviewer: ```suggestion
developers do not need to worry about key collisions, but the implementation of Substrate Storage Maps does become
```

  [`StorageMap#mutate(key, fn)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#tymethod.mutate),
	[`StorageDoubleMap#mutate(key1, key2, fn)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#tymethod.mutate)
* `take` - Remove the value associated with the given key from storage. Docs:
  [`StorageMap#take(key)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#tymethod.take),
	[`StorageDoubleMap#take(key1, key2)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#tymethod.take)

### Iterable Storage Maps

The Substrate storage API provides iterable map implementations. Because maps are often used to track unbounded sets of
data (account balances, for example) it is especially likely to exceed block production time by iterating over maps in
their entirety within the runtime. Furthermore, because maps are comprised of more layers of indirection than native
lists, they are significantly more costly than lists to iterate over with respect to time. This is not to say that it

> is "wrong" to iterate over maps in your runtime; in general Substrate focuses on "first principals" as opposed to hard
> and fast rules of right and wrong. Being efficient within the runtime of a blockchain is an important first principal

Reviewer: This one is definitely principle
```suggestion
is "wrong" to iterate over maps in your runtime; in general Substrate focuses on "first principles" as opposed to hard
and fast rules of right and wrong. Being efficient within the runtime of a blockchain is an important first principle
```

of Substrate and this information is designed to help you understand _all_ of Substrate's storage capabilities and use

> them in a way that respects the important first principals around which they were designed. Depending on

Reviewer: ```suggestion
them in a way that respects the important first principles around which they were designed. Depending on
```

## Storage Items

[The `storage` module in the Substrate Support pallet](https://substrate.dev/rustdocs/master/frame_support/storage/index.html)
gives runtime developers access to Substrate's flexible storage APIs. Any value which can be encoded by the
[Parity SCALE codec](../advanced/codec) is supported by these storage APIs:

* [Storage Value](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html) - A single value
* [Storage Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html) - A key-value hash map
* [Storage Double Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html) - An
  implementation of a map with two keys that provides the important ability to efficiently remove all entries that have
  a common first key
* Iterable Storage Maps - Implementations of
  [Storage Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageMap.html) and
  [Storage Double Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageDoubleMap.html)

>   whose keys and values can be iterated over

Reviewer 2: I would have actually thought that we would present 3 kind of storage: value, map and double_map and map and double_map are iterable when they use hasher with key information.

like in my mind they declare a map storage and this map storage implements StorageMap and optionnally IterableStorageMap traits.

I think it is less confusing presented this way, but I have no strong opinion

Author: I would be interested to hear the opinions of other people on this topic. I am trying to approach this document from the POV of the capabilities that users will be consuming, which is why I have chosen to describe things the way that I do.

Reviewer 3: To our users, there are 3 types of storage: value, map, double_map.

The traits on top of these storage are opaque to the user, so I agree with @thiolliere

The type of Storage Item you select should depend on the logical way in which the value will be used by your runtime.

### Storage Value

This type of storage item should be used for values that are viewed as a single unit by the runtime, whether that is a
single primitive value, a single `struct` or a single collection of related items. Although wrapping related items in
a shared `struct` is an excellent way to reduce the number of storage reads (an important consideration), at some point
the size of the object will begin to incur costs that may outweigh the optimization in storage reads. Storage values
can be used to store lists of items, but runtime developers should take care with respect to the size of these lists.
Large lists incur storage costs just like large `structs`. Furthermore, iterating over a large list in your runtime may

> result in exceeding the block production time.

Reviewer 2: Yes but I guess if your list is too long to be iterated upon, then just decoding it would be already very very long. I think the emphasize on the cost to read/write it is more important. but ok anyway

Author: I believe that Shawn has indicated we should specifically caution people against iteration within the runtime.

Reviewer 2: (ok but just to clarify if you store in your list something different from basic u32, u64... then decoding/encoding is done by iterating other elements.)

Author: Can you please expand on that concept just a bit for me? Are you referring to the fact that anything other than a Rust primitive requires a pointer to be dereferenced?

Reviewer 2: in parity-scale-codec to decode `Vec<u32>` and all other type in `	U8, I8, U16, I16, U32, I32, U64, I64, U128, I128,` we decode/encode by doing mem::transmute, otherwise we do:
```rust
fn encode(..) {
...
				for item in self {
					item.encode_to(dest);
				}
}
```
and 
```rust
fn decode(...) {
...
					for _ in 0..len {
						r.push(T::decode(input)?);
					}
}
```

Author: I see, so complex data types require an extra layer of indirection that materializes as an encoding/decoding step. Correct?

Reviewer 2: I wouldn't phrases this way. Just encoding/decoding of vec is optimized for `U8, I8, U16, I16, U32, I32, U64, I64, U128, I128` I think also Option is optimized for them and some other generic data type.
But this is just optimization in my point of view.

Author: Do you think that it is important for us to document this optimization with respect to the runtime storage interface?

Reviewer 2: not sure, should belong to some codec doc I think

Author: Opened a separate issue :+1: #43

#### Methods

Refer to the Storage Value documentation for
[a comprehensive list of the methods that Storage Values expose](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#required-methods).
Some of the most important methods are summarized here:

##### `get()`

Load the value from storage.

##### `put(val)`

Store the provided value.

##### `mutate(fn)`

Mutate the value with the provided function.

##### `take()`

> Remove the value from storage.

Reviewer 2: maybe the name of those method are self explanatory and should just directly link to the inner doc? (not sure)

Author: I tried to take a best-of-both-worlds approach by deemphasizing these sections within the document. I think it's good context to provide, but agree that the Rust docs should be the canonical source with respect to this kind of stuff.

### Storage Maps

> Storage Maps are implemented as hash maps, which is a pattern that should be familiar to most developers. In order to

Reviewer 2: > Storage Maps are implemented as hash maps

maybe it is only me but for me hash map is a data structure which handle collision. at least if presented this way it should be clear that collision are not handled.

But actually instead of using "hash map" term I would rather say that it is a map where key are hashed.

Author: Thank you for this clarification. Please let me know if you think my solution was sufficient.

Reviewer 2: hmm I see, I'm not sure actually what I proposed was better, like

> Storage Maps are implemented as maps with hashed keys, which is a pattern that should be familiar to most developers.

wouldn't have make sense to me I think if I didn't know it already, not sure how to do better. (still for me the clearest is that a storage map insert at the path `hash(encode(key))` the value encoded.)

Author: Tried to clarify this in the docs :+1:

give blockchain engineers increased control over the way in which these data structures are stored, Substrate allows
developers to select the hashing algorithm that is used to generate map keys. Map data structures are ideal for managing

> sets of items whose elements will be accessed randomly, as opposed to iterating over them sequentially in their entirety.

Reviewer 2: I think it would be better to show how it is implemented in the underlying trie to make it more understandable about advantage/disadvantage.

Basically the value is put in the trie at the path==some_storage_prefix++hashed_key.
So each value is stored in an individual node, so reading/writing a value only decode/encode one raw_value which is stored in one trie-node.

also the performance of the map will depends on how big the map is because read/writing one value will recompute the path==some_prefix++hashed_key. and the hashed_key part can be more or less costy to compute. https://www.shawntabrizi.com/substrate/transparent-keys-in-substrate/

Author: There is [a separate document that goes over how storage is implemented](https://github.com/substrate-developer-hub/knowledgebase/blob/runtime-storage/current/advanced/storage.md). This document is meant to explain Substrate's storage interfaces and help blockchain runtime developers understand when and how to use them.

Reviewer 2: this document will not explain how frame_support::storage abstraction are implemented in storage do they ? so I would explain their implementation here actually.

Author: I believe the purpose of the advance document is to cover implementation and the purpose of this document (the runtime document) is to cover the interface. @joepetrowski can you provide your input?

#### Methods

[Storage Maps expose an API](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#required-methods)
that is similar to that of Storage Values. The selected methods referenced below all use a single key; use two keys for

> [Storage Double Maps](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#required-methods):

Reviewer 2: Here also I think we should say that using this two keys result in this in the trie: key1, key2, value is stored at
value is stored at some_prefix++hashe1(key1)++hasher2(key2).
and so you can iterate on one key1.

Author: This is information I also think would belong in the advanced document.

##### `get(key)`

Load the value associated with the provided key from storage.

##### `store(key, val)`

Store the provided value by associating it with the given key.

##### `mutate(key, fn)`

Use the provided function to mutate the value associated with the given key.

##### `take(key)`

Remove the value associated with the given key from storage.

### Iterable Storage Maps

The Substrate storage API provides iterable map implementations. Because maps are often used to track unbounded sets of
data (account balances, for example) it is especially likely to exceed block production time by iterating over maps in
their entirety within the runtime. Furthermore, because maps are comprised of more layers of indirection than native

> lists, they are significantly more costly than lists to iterate over with respect to time. Depending on

Reviewer 2: I think we can just explain what is the "more layers of indirection than native lists", basically iterating over a map is just iterating over all keys after the prefix `storage_prefix` so this trie iteration is more costy than native list because it is a trie.

Author: I felt like that was a step too far for this document and that topics like that belong in [the advanced storage documentation](https://github.com/substrate-developer-hub/knowledgebase/blob/runtime-storage/current/advanced/storage.md).

Reviewer 2: The above document just explain how the trie works not how frame_support storage abstractions are implemented using the trie, no ?
As a user in order to understand more precisely the cost of those abstraction, explaining their implementation on the trie seems useful (maybe it belong somewhere else in the doc but still I feel like saying that a storage map just store all its value after a prefix and iterating upon means iterating the trie after a prefix is quite straightforward and quickly understandable)

> Because maps are often used to track unbounded sets of
data it is especially likely to exceed block production time by iterating over maps in
their entirety within the runtime

This is just a trade off, some storage map in the runtime are quite bounded actually. But we use storage map because they are more often accessed by individual value than as a whole. Thus using a storage map is more efficient because accessing one value is cheaper, but it is sometime iterated upon still because it is not too big.

Author: Tried to clarify this in the docs :+1:

[the hashing algorithm](#Transparent-Hashing-Algorithms) that you select to generate a map's keys, you may be able to

> iterate across its keys as well as its values.

Reviewer 2: Maybe we should presice here that IterableStorageMap is only implemented for Map if the hasher is reversible. (i.e. implements ReversibleStorageHasher, which is not shown in the doc but should be IMO)

Author: Is `ReversibleStorageHasher` documented anywhere in the Rust docs? I only see it mentioned in [the docs for `IterableStorage(Double)Map`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageMap.html#implementors) but it's not indexed so difficult to link to.

Reviewer 2: it is in the doc since https://github.com/paritytech/substrate/pull/5504 should be soon in the generated doc I expect. I agree it was a probably mistake not to expose it.

Reviewer 2: but thinking again it seems that non reversible hasher are deprecated so maybe not worth mentionning:
```
/// Deprecated hashers, which do not support iteration over keys include:
/// * `twox_128` - TwoX with 128bit.
/// * `twox_256` - TwoX with with 256bit.
/// * `blake2_128` - Blake2 with 128bit.
/// * `blake2_256` - Blake2 with 256bit.
```

#### Methods

[Iterable Storage Maps expose the following methods](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageMap.html#required-methods)

> in addition to the other map methods:

Reviewer 2: their method are slightly different for map and doublemap

Author: Tried to clarify this in the docs :+1:

##### `iter()`

Enumerate all elements in the map in no particular order. If you alter the map while doing this, you'll get undefined
results.

##### `drain()`

Remove all elements from the map and iterate through them in no particular order. If you add elements to the map while
doing this, you'll get undefined results.

##### `translate(fn)`

Use the provided function to translate all elements of the map, in no particular order. To remove an element from the

> map, return `None` from the translation function.

Reviewer 2: they also differ between map and double_map (as a note I tried to unify interface here https://github.com/paritytech/substrate/pull/5335)

Author: Tried to clarify this in the docs :+1:

Large lists incur storage costs just like large `structs`. Furthermore, iterating over a large list
in your runtime may result in exceeding the block production time - if this occurs your blockchain
will stop producing blocks, which means that it will stop functioning.

#### Methods

Refer to the Storage Value documentation for
[a comprehensive list of the methods that Storage Values expose](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#required-methods).
Some of the most important methods are summarized here:

- [`get()`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#tymethod.get) -
  Load the value from storage.
- [`put(val)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#tymethod.put) -
  Store the provided value.
- [`mutate(fn)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#tymethod.mutate) -
  Mutate the value with the provided function.
- [`take()`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#tymethod.take) -

>   Remove the value from storage.

Reviewer: ```suggestion
  Load the value and remove it from storage.
```

that is similar to that of Storage Values.

- `get` - Load the value associated with the provided key from storage. Docs:
  [`StorageMap#get(key)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#tymethod.get),
  [`StorageDoubleMap#get(key1, key2)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#tymethod.get)
- `insert` - Store the provided value by associating it with the given key. Docs:
  [`StorageMap#insert(key, val)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#tymethod.insert),
  [`StorageDoubleMap#insert(key1, key2, val)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#tymethod.insert)
- `mutate` - Use the provided function to mutate the value associated with the given key. Docs:
  [`StorageMap#mutate(key, fn)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#tymethod.mutate),
  [`StorageDoubleMap#mutate(key1, key2, fn)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#tymethod.mutate)

> - `take` - Remove the value associated with the given key from storage. Docs:

Reviewer: ```suggestion
- `take` - Load the value associated with the given key and remove it from storage. Docs:
```

  [`StorageMap#take(key)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#tymethod.take),
  [`StorageDoubleMap#take(key1, key2)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#tymethod.take)

#### Iterable Storage Maps

Depending on [the hashing algorithm](#Transparent-Hashing-Algorithms) that you select to generate a
map's keys, you may be able to iterate across its keys and values. Because maps are often used to
track unbounded sets of data (account balances, for example) it is especially likely to exceed block
production time by iterating over maps in their entirety within the runtime. Furthermore, because

> maps are comprised of more layers of indirection than native lists, they are significantly more

Reviewer: What are layers of indirection?

  doing this, you'll get undefined results. Docs:
  [IterableStorageMap#iter()](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageMap.html#tymethod.iter),
  [IterableStorageDoubleMap#iter(key1)](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageDoubleMap.html#tymethod.iter)
- `drain` - Remove all elements from the map and iterate through them in no particular order. If you
  add elements to the map while doing this, you'll get undefined results. Docs:
  [IterableStorageMap#drain()](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageMap.html#tymethod.drain),
  [IterableStorageDoubleMap#drain(key1)](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageDoubleMap.html#tymethod.drain)
- `translate` - Use the provided function to translate all elements of the map, in no particular
  order. To remove an element from the map, return `None` from the translation function. Docs:
  [IterableStorageMap#translate(fn)](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageMap.html#tymethod.translate),

>   [IterableStorageDoubleMap#translate(fn)](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageDoubleMap.html#tymethod.translate)

Reviewer: Should have \` \` around the linked items.

This table lists some common hashers used in Substrate and denotes those that are cryptographic and
those that are transparent:

| Hasher                                                                                                | Cryptographic | Transparent |
| ----------------------------------------------------------------------------------------------------- | ------------- | ----------- |
| [Blake2 128](https://substrate.dev/rustdocs/master/frame_support/struct.Blake2_128.html)              | X             |             |
| [TwoX 128](https://substrate.dev/rustdocs/master/frame_support/struct.Twox128.html)                   |               |             |
| [Blake2 128 Concat](https://substrate.dev/rustdocs/master/frame_support/struct.Blake2_128Concat.html) | X             | X           |
| [TwoX 64 Concat](https://substrate.dev/rustdocs/master/frame_support/struct.Twox64Concat.html)        |               | X           |
| [Identity](https://substrate.dev/rustdocs/master/frame_support/struct.Identity.html)                  |               |             |

The Identity hasher encapsulates a hashing algorithm that has an output equal to its input (the

> identity function).This type of hasher should only be used when the starting key is already a

Reviewer: ```suggestion
identity function). This type of hasher should only be used when the starting key is already a
```

cryptographic hash.

## Declaring Storage Items

You can use
[the `decl_storage` macro](https://substrate.dev/rustdocs/master/frame_support/macro.decl_storage.html)
to easily create new runtime storage items. Here is an example of what it looks like to declare each
type of storage item:

```rust
decl_storage! {
	trait Store for Module<T: Trait> as Example {
		pub SomePrimitiveValue get(fn some_primitive_value): u32;
		// complex types are prefaced by T::
		pub SomeComplexValue: T::AccountId;

> 		pub SomeMap get(fn some_map): map hasher(blake2_128_concat) str => u32;

Reviewer: We recommend against strings in the runtime, so I would change this type.

Reviewer: Could also add in a private storage item and explain its accessibility.

- `mutate` - Use the provided function to mutate the value associated with the given key. Docs:
  [`StorageMap#mutate(key, fn)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#tymethod.mutate),
  [`StorageDoubleMap#mutate(key1, key2, fn)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#tymethod.mutate)
- `take` - Load the value associated with the given key and remove it from storage. Docs:
  [`StorageMap#take(key)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#tymethod.take),
  [`StorageDoubleMap#take(key1, key2)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#tymethod.take)

#### Iterable Storage Maps

Depending on [the hashing algorithm](#Transparent-Hashing-Algorithms) that you select to generate a
map's keys, you may be able to iterate across its keys and values. Because maps are often used to
track unbounded sets of data (account balances, for example) it is especially likely to exceed block

> production time by iterating over maps in their entirety within the runtime. Furthermore, because

Reviewer 4: > Furthermore, because
accessing the elements of a map requires more pointer dereferencing than accessing the elements of a
native list, maps are significantly _more_ costly than lists to iterate over with respect to time.

Can you explain this a bit more? I am not sure where it comes from and until now my understanding of the state was that the most important factors encoding + db access. In-memory dereferencing seems negligible to me.

Reviewer 4: (I agree that iterating maps is expensive, I am not sure if your reason is correct.)

Author: So, this line used to read as follows:
> Furthermore, because maps are comprised of more layers of indirection than native lists...

I agree with you that the change didn't necessarily make things more clear. I guess what I meant above was that you don't need to dereference _memory_ pointers, per se, but you do have to dereference the pointer to the element in the map...I think I'm trying to describe the "db access" part that you mention above. I will think about alternate ways to phrase this but would definitely appreciate any suggestions :+1:

Author: Okay, this now reads:
> ...accessing the elements of a map requires more database reads than accessing the elements of a native list...

I hope you agree the simplest approach was the best one in this case :sweat_smile:

accessing the elements of a map requires more pointer dereferencing than accessing the elements of a
native list, maps are significantly _more_ costly than lists to iterate over with respect to time.
This is not to say that it is "wrong" to iterate over maps in your runtime; in general Substrate
focuses on "[first principles](#Best-Practices)" as opposed to hard and fast rules of right and
wrong. Being efficient within the runtime of a blockchain is an important first principle of
Substrate and this information is designed to help you understand _all_ of Substrate's storage
capabilities and use them in a way that respects the important first principles around which they

> were designed.

Reviewer 4: would be nice to mention here when and which maps are iterable at the end of the day.

Author: This is a great point and I'm starting to wonder if we should even document non-transparent hashers since they are now deprecated. Should we just behave as if all hashers are transparent and all maps are iterable?

Author: Asked @shawntabrizi about this offline and he agrees that we should take the simpler approach: only document that which is not deprecated :+1:

Author: Tried to clarify this in the docs :+1:

##### Cryptographic Hashing Algorithms

Cryptographic hashing algorithms are those that use cryptography to make it challenging to use the
input to the hashing algorithm to influence its output. For example, a cryptographic hashing
algorithm would produce a wide distribution of outputs even if the inputs were the numbers 1
through 10. It is critical to use cryptographic hashing algorithms when users are able to influence
the keys of a Storage Map. Failure to do so creates an attack vector that makes it easy for
malicious actors to degrade the performance of your blockchain network. An example of a map that
should use a cryptographic hash algorithm to generate its keys is a map used to track account
balances. In this case, it is important to use a cryptographic hashing algorithm so that an attacker
cannot bombard your system with many small transfers to sequential account numbers; without a
cryptographic hash algorithm this would create an imbalanced storage structure that would suffer in
performance. Cryptographic hashing algorithms are more complex and resource-intensive than their
non-cryptographic counterparts, which is why Substrate allows developers to select when they are

> used.

Reviewer 4: would be good to name the most common hashes here and say which is secure and which is not.

identity function). This type of hasher should only be used when the starting key is already a
cryptographic hash.

## Declaring Storage Items

You can use
[the `decl_storage` macro](https://substrate.dev/rustdocs/master/frame_support/macro.decl_storage.html)
to easily create new runtime storage items. Here is an example of what it looks like to declare each
type of storage item:

```rust
decl_storage! {
	trait Store for Module<T: Trait> as Example {
		SomePrivateValue: u32;
		pub SomePrimitiveValue get(fn some_primitive_value): u32;
		// complex types are prefaced by T::
		pub SomeComplexValue: T::AccountId;
		pub SomeMap get(fn some_map): map hasher(blake2_128_concat) T::AccountId => u32;
		pub SomeDoubleMap: double_map hasher(blake2_128_concat) u32, hasher(blake2_128_concat) T::AccountId => u32;
	}
}
```

Notice that the map storage items specify [the hashing algorithm](#Hashing-Algorithms) that will be
used.

### Visibility

In the example above, all the storage items except `SomePrivateValue` are made public by way of the
`pub` keyword. Blockchain storage is always publicly
[visible from _outside_ of the runtime](#Accessing-Storage-Items); the visibility of Substrate

> storage items only impacts whether or not other runtime pallets will be able to access the storage

Reviewer 4: ```suggestion
storage items only impacts whether or not other pallets _within the runtime_ will be able to access the storage
```

transparent hashing algorithm if you would like access [iterable map](#Iterable-Storage-Maps)
capabilities. Refer to [the advanced storage documentation](../advanced/storage#Storage-Map-Keys) to
learn more about the important capabilities that transparent hashing algorithms expose.

##### Common Substrate Hashers

This table lists some common hashers used in Substrate and denotes those that are cryptographic and
those that are transparent:

| Hasher                                                                                     | Cryptographic | Transparent |
| ------------------------------------------------------------------------------------------ | ------------- | ----------- |
| [Blake2 128 Concat](https://crates.parity.io/frame_support/struct.Blake2_128Concat.html)   | X             | X           |
| [TwoX 64 Concat](https://crates.parity.io/frame_support/struct.Twox64Concat.html)          |               | X           |
| [Identity](https://crates.parity.io/frame_support/struct.Identity.html)                    |               |             |
| [Blake2 128](https://crates.parity.io/frame_support/struct.Blake2_128.html) **DEPRECATED** | X             |             |

> | [TwoX 128](https://crates.parity.io/frame_support/struct.Twox128.html) **DEPRECATED**      |               |             |

Reviewer 2: there is actually 4 depracated hasher but I don't mind if we don't mention them:

 >   twox_128 - TwoX with 128bit.
    twox_256 - TwoX with with 256bit.
    blake2_128 - Blake2 with 128bit.
    blake2_256 - Blake2 with 256bit.

The Identity hasher encapsulates a hashing algorithm that has an output equal to its input (the
identity function). This type of hasher should only be used when the starting key is already a
cryptographic hash.

## Declaring Storage Items

You can use
[the `decl_storage` macro](https://crates.parity.io/frame_support/macro.decl_storage.html) to easily
create new runtime storage items. Here is an example of what it looks like to declare each type of
storage item:

```rust
decl_storage! {
	trait Store for Module<T: Trait> as Example {
		SomePrivateValue: u32;
		pub SomePrimitiveValue get(fn some_primitive_value): u32;

> 		// complex types are prefaced by T::
> 		pub SomeComplexValue: T::AccountId;

Reviewer 2: I feel like the wording shows that there is macro magic in the syntax here. but actually this is a regular rust type.
(The macro parse the type themself to know if there are generic over `T`, if they are then the storage struct generated is generic as well.)

```suggestion
		// types can make use of generic `T: Trait`
		pub SomeComplexValue: T::AccountId;
```

}
```

### Genesis Config

You can define
[an optional `GenesisConfig`](https://substrate.dev/rustdocs/master/frame_support/macro.decl_storage.html#genesisconfig)
struct in order to initialize Storage Items in the genesis block of your blockchain.

// TODO

## Accessing Storage Items

Blockchains that are built with Substrate expose a remote procedure call (RPC) server that can be
used to query your blockchain's runtime storage. You can use software libraries like
[Polkadot JS](https://polkadot.js.org/) to easily interact with the RPC server from your code and
access storage items. The Polkadot JS team also maintains
[the Polkadot Apps UI](https://polkadot.js.org/apps), which is a fully-featured web app for
interacting with Substrate-based blockchains, including querying storage. Refer to
[the advanced storage documentation](../advanced/storage) to learn more about how Substrate uses a
key-value database to implement the different kinds of Storage Items and how to query this database
directly by way of the RPC server.

## Child Storage Tries

TODO

## Storage Cache

TODO

> ## Best Practices

Reviewer: Let's add:

- Verify first, write last
- Store only consensus critical info on chain (use example of submitting preimage)
- Set limits to the length of vectors that users can affect

}
```

### Genesis Config

You can define
[an optional `GenesisConfig`](https://substrate.dev/rustdocs/master/frame_support/macro.decl_storage.html#genesisconfig)
struct in order to initialize Storage Items in the genesis block of your blockchain.

> // TODO

Reviewer: @thiolliere what info do you think should go here?

Reviewer 2: GenesisConfig is generated by decl_storage macro, it does initialize storage items for the genesis block of the blockchain.

Its logic is defined by:
* `config` and `build` attribute when declaring an individual storage.
* add_extra_genesis informations. which allows to add fields in the GenesisConfig struct and a build function to build storages using those fields.

(I can be more precise actually we execute first for each storage the build closure if it exist, or if no build attribute but a config attribute then we add into storage the value found in config. Then we execute the add_extra_genesis::build function.)

governance mechanisms allow this to be done in a way that defers bringing all the data associated
with a proposal on-chain until _after_ it has been approved. This means that storage is not wasted
on proposals that fail. Once a proposal has passed, someone can initiate the actual dispatchable
call (including all its parameters), which will be hashed and compared to the hash in the proposal.
Another common pattern for using hashes to minimize data that is stored on-chain is to store the
metadata associated with an object in [IPFS](https://ipfs.io/); this means that only the IPFS
location (a hash that is bounded in size) needs to be stored on-chain.

Hashes are only one mechanism that can be used to control the size of runtime storage. An example of
another mechanism is [bounds](#Create-Bounds).

### Verify First, Write Last

The state of a blockchain network's storage is immutable; data can be changed, but there will always
be a record of these changes, and making them typically incurs costs. Because of this, it is
important that data is only persisted to runtime storage when it is certain that all preconditions

> have been met. In general, code blocks that may result in adding data to storage should be

Reviewer: ```suggestion
have been met. In general, code blocks that may result in mutating storage should be
```

been met.

### Create Bounds

Creating bounds on the size of storage items is an extremely effective way to control the use of
runtime storage and one that is used repeatedly throughout the Substrate codebase. In general, any
storage item whose size is determined by user action should have a bound on it.
[The multi-signature capabilities from the Utility pallet](https://crates.parity.io/pallet_utility/trait.Trait.html#associatedtype.MaxSignatories)
that were described above are one such example. In this case, the list of signatories associated
with a multi-signature operation is provided by the multi-signature participants. Because this
signatory list is [necessary to come to consensus](#What-to-Store) on the state of the
multi-signature operation, it must be stored in the runtime. However, in order to give runtime
developers control over how much space in storage these lists may occupy, the Utility pallet
requires users to configure a bound on this number that will be included as a
[precondition](#Verify-First-Write-Last) before anything is written to storage.

## Child Storage Tries

TODO

## Storage Cache

TODO

> 

Reviewer: This may be good in the Advanced doc but don't think it's needed here. TBH I'm not sure how much runtime developers even need to know about these two things.

}
```

### Genesis Configuration

Substrate's runtime storage APIs include capabilities to initialize storage items in the genesis
block of your blockchain. The genesis storage configuration APIs expose a number of mechanisms for
initializing storage, all of which have entry points in the `decl_storage` macro. These mechanisms
all result in the creation of a `GenesisConfig` data type that implements
[the `sp_runtime::BuildModuleGenesisStorage` trait](https://crates.parity.io/sp_runtime/trait.BuildModuleGenesisStorage.html)

> and will be added to the storage item's module (e.g.

Reviewer 2: I'm not sure I read this correctly, what does the runtime GenesisConfig being added to the storage item's module means ?

Maybe we could make it more clear that there is a GenesisConfig overarching type for the runtime which is made by construct_runtime by joining the specified pallet genesis config (specified with `Config`)

storage items that are tagged for genesis configuration will have a corresponding attribute on this
data type. In order to consume a module's genesis configuration capabilities, you must include the
`Config` element when adding the module to your runtime with
[the `construct_runtime` macro](https://crates.parity.io/frame_support/macro.construct_runtime.html).
All the `GenesisConfig` types for the modules that inform a runtime will be aggregated into a single
`GenesisConfig` type for that runtime, which implements
[the `sp_runtime::BuildStorage` trait](https://crates.parity.io/sp_runtime/trait.BuildStorage.html)
(e.g.
[`Struct node_template_runtime::GenesisConfig`](https://crates.parity.io/node_template_runtime/struct.GenesisConfig.html));
each attribute on this type corresponds to a `GenesisConfig` from one of the runtime's modules.
Ultimately, the runtime's `GenesisConfig` is exposed by way of
[the `sc_chain_spec::ChainSpec` trait](https://crates.parity.io/sc_chain_spec/trait.ChainSpec.html).
For a complete and concrete example of using Substrate's genesis storage configuration capabilities,
refer to the `decl_storage` macro in

> [the Society pallet](https://github.com/paritytech/substrate/blob/master/frame/society/src/lib.rs)

Reviewer 2: should it be better to point to example pallet ?

Author: I don't believe that the example pallet does any genesis configuration.

on proposals that fail. Once a proposal has passed, someone can initiate the actual dispatchable
call (including all its parameters), which will be hashed and compared to the hash in the proposal.
Another common pattern for using hashes to minimize data that is stored on-chain is to store the
metadata associated with an object in [IPFS](https://ipfs.io/); this means that only the IPFS
location (a hash that is bounded in size) needs to be stored on-chain.

Hashes are only one mechanism that can be used to control the size of runtime storage. An example of
another mechanism is [bounds](#Create-Bounds).

### Verify First, Write Last

The state of a blockchain network's storage is immutable; data can be changed, but there will always
be a record of these changes, and making them typically incurs costs. Because of this, it is
important that data is only persisted to runtime storage when it is certain that all preconditions
have been met. In general, code blocks that may result in mutating storage should be structured as
follows:

```rust
{
  // all checks and throwing code go here

  // all storage writes go here; no throwing code below this line

>   // all event emissions go here

Reviewer 2: events can be interleaved with storage writes

### References

* Visit the reference docs for the

>   [`decl_storage!` macro](https://substrate.dev/rustdocs/master/frame_support/macro.decl_storage.html) more details

Reviewer: ```suggestion
  [`decl_storage!` macro](https://substrate.dev/rustdocs/master/frame_support/macro.decl_storage.html) for more details about
```
